File: code/213.go
/*
 * @Author: lmio
 * @Date: 2023-04-02 15:30:36
 * @LastEditTime: 2023-04-07 14:35:57
 * @FilePath: /leetcode/code/213.go
 * @Description:213. 打家劫舍 II
 */
package code

import "leetcode/utils"

func RobII(nums []int) int {
	// 返回不偷窃第一家与不偷窃最后一家的最大值
	return utils.Max(RobI(nums[1:]), RobI(nums[:len(nums)-1]))
}

File: code/213.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-09-17 18:53:18
LastEditors: lmio 2091319361@qq.com
Description: 213. 打家劫舍 II
'''

from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        def f(arr: List[int]) -> int:
            if not arr:
                return 0
            n = len(arr)
            if n == 1:
                return arr[0]
            dp = [0]*n
            dp[0] = arr[0]
            dp[1] = max(arr[0], arr[1])
            for i in range(2, n):
                dp[i] = max(dp[i-2]+arr[i], dp[i-1])
            return dp[-1]
        return max(f(nums[1:]), f(nums[:len(nums)-1]))